### PR TITLE
[JENKINS-9104] Add process killing whitelist

### PR DESCRIPTION
**Pull request by DanielWeber**
In order to let daemon processes like mspdbsrv.exe survive after-build process cleanup, people
can now configure a list of process names which should not be killed.

### CODE REVIEW

**Comment by daniel-beck** [comment]
Please revise your commit to remove the unnecessary whitespace changes, reordering of imports, and reformatting of code from your pull request.

**Comment by DanielWeber** [comment]
I amended the commit to avoid reordering imports and whitespace changes.

File: core/src/main/java/jenkins/model/GlobalProcessWhitelistConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2011, CloudBees, Inc.
```
Comment (KostyaSha) [inline core/src/main/java/jenkins/model/GlobalProcessWhitelistConfiguration.java]: If it new, then why 2011 and are you from CB?

**Comment by KostyaSha** [comment]
How it differs from https://wiki.jenkins-ci.org/display/JENKINS/Spawning+processes+from+build ?

**Comment by oleg-nenashev** [comment]
* The change won't work properly on Windows. See https://issues.jenkins-ci.org/browse/JENKINS-19156 and https://issues.jenkins-ci.org/browse/JENKINS-20289
* Seems the change introduces massive binary compatibility issues, which may affect plugins and other API users
* I'm not sure why it's required (see comments from @KostyaSha )

**Comment by DanielWeber** [comment]
@KostyaSha : There's a long discussion in [JENKINS-9104](https://issues.jenkins-ci.org/browse/JENKINS-9104). People should not be forced to disable process killing for complete jobs or even the whole Jenkins instance. Starting mspdbsrv manually is also unreliable, see [my comment](https://issues.jenkins-ci.org/browse/JENKINS-9104?focusedCommentId=162154&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-162154) and the [follow up](https://issues.jenkins-ci.org/browse/JENKINS-9104?focusedCommentId=162171&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-162171). Steve Carter added a [summary comment](https://issues.jenkins-ci.org/browse/JENKINS-9104?focusedCommentId=220953&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-220953) recently.

File: core/src/main/java/jenkins/model/GlobalProcessWhitelistConfiguration.java (hunk 1 above)
Comment (DanielWeber) [inline core/src/main/java/jenkins/model/GlobalProcessWhitelistConfiguration.java]: Sorry, my mistake. Was copy'n'pasted.

**Comment by DanielWeber** [comment]
@oleg-nenashev I don't see how the issues you mentioned affect the whitelisting change. The ProcessKillers also won't be called if a process is whitelisted.

Regarding binary compatibility issues: I'd have to find a way to do this without interface changes. I just wanted to avoid ProcessTree reaching out to Jenkins.getInstance() to get the whitelist. Seemed kind of ugly.

See my previous comment for why I think we still need this.

**Comment by DanielWeber** [comment]
To all: Thanks for your feedback!

**Comment by daniel-beck** [comment]
Maybe this could be an extension point and you use `ExtensionList` to get it. That way, MSBuild plugin could register `mspdbsrv.exe` to be whitelisted. A new plugin could add the UI around a user-configurable whitelist.

Would this make sense?

File: core/src/main/java/jenkins/model/GlobalProcessWhitelistConfiguration.java (hunk 1 above)
Comment (DanielWeber) [inline core/src/main/java/jenkins/model/GlobalProcessWhitelistConfiguration.java]: Fixed

File: core/src/main/java/hudson/Launcher.java [hunk 2]
```diff
@@ -889,7 +907,9 @@ public Channel launchChannel(String[] cmd, OutputStream out, FilePath workDir, M
         }
 
         @Override
-        public void kill(Map<String, String> modelEnvVars) throws IOException, InterruptedException {
+        public void kill(Map<String, String> modelEnvVars,
+                List<String> whitelist) throws IOException,
```
Comment (KostyaSha) [inline core/src/main/java/hudson/Launcher.java]: imho, in this case arguments on one line and throws on other

**Comment by oleg-nenashev** [comment]
@DanielWeber 

>  I don't see how the issues you mentioned affect the whitelisting change. The ProcessKillers also won't be called if a process is whitelisted.

It will check top-level processes only. E.g. a user won't be able to use whitelist for processes he starts from the "Execute Batch" step. BTW, it's not an obstacle for your change, just a limitation due to the existing issue

 > I just wanted to avoid ProcessTree reaching out to Jenkins.getInstance() to get the whitelist. 

It worth to mention that <code>Jenkins::getInstance()</code> may return null when Jenkins starts up or shuts down. In such case the configs may be unavailable

@daniel-beck 
:+1: for the extension point. A raw process list implementation may be left within the core to provide a basic functionality.

**Comment by DanielWeber** [comment]
@daniel-beck I also like the extension point idea, in this case I would suggest to provide a "process killing veto" extension point. So instead of only supporting whiltelisting by name, plugins could veto process killing based on other criteria as well. And MSBuild plugin users won't even have to know or think about this :)

**Comment by ki82** [comment]
Is there anything that blocks this from being merged?

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> A raw process list implementation may be left within the core to provide a basic functionality.

I don't think this is needed. Like e.g. Secure Requester Whitelist Plugin, this can be done separately. In general, I doubt many users will need this feature to be customizable. I've only really seen issues with mspdbsrv, everyone else can deal with killing or setting `BUILD_ID`.

@ki82 

> Is there anything that blocks this from being merged?

You mean besides the three recent comments favoring a different solution to this problem?

**Comment by DanielWeber** [comment]
I created a new PR, adding the extension point to Jenkins core as discussed above, see https://github.com/jenkinsci/jenkins/pull/1684

**Comment by TheFriendlyCoder** [comment]
Hello everyone. I am just looking to get an update on the progress of this work. We have been avoiding upgrading our Jenkins instance for over a year now because of this one specific bug, and it appears there has been some progress on fixing it in recent weeks, however the latest LTS release still does not appear to have the full fix incorporated into it.

Are there any plans to finish incorporating this PR into an LTS release soon or do we have to start looking at one of the many hacks described under JENKINS-9104 instead?

**Comment by oleg-nenashev** [comment]
@TheFriendlyCoder 
It's a new feature. I suppose it's not a subject for the backporting, so it may appear in the next LTS line only

**Comment by olivergondza** [comment]
@TheFriendlyCoder, nothing will be incorporated into LTS until it is reliably fixed on master. If there is something fixed you think should go to LTS, please [nominate a candidate](https://wiki.jenkins-ci.org/display/JENKINS/LTS+Release+Line).

**Comment by TheFriendlyCoder** [comment]
@oleg-nenashev 
The defect in question which instigated this pull request was a defect reported on the current LTS edition. In fact, the defect was introduced in a maintenance release for the current LTS edition some time ago, so I would truly hope the fix would be backported there as well.

**Comment by TheFriendlyCoder** [comment]
@olivergondza 
Is it typical that bugs reported on the LTS version be fixed only on the master? I would have expected them to be fixed on the LTS version as well, if not first.

In the case where a bug is reported against an LTS edition, and fixed on the master, is it still required to nominate it for backporting? That seems a bit odd to me. Shouldn't such bugs automatically be backported by convention?

**Comment by jtnord** [comment]
@TheFriendlyCoder no.
If bugs where fixed on stable branches first this could introduce more regressions as they will not have been exposed to users testing them.  This is why bugs are always fixed on master first.

For a fix to be accepted into an LTS release it generally needs to have soaked in the master branch for a few weeks as well (excluding security fixes). some details [here](https://wiki.jenkins-ci.org/display/JENKINS/LTS+Release+Line#LTSReleaseLine-Backportingprocess)

**Comment by olivergondza** [comment]
@TheFriendlyCoder 
Unfortunately, there is no structural information in jira pointing to a version where the bug was spotted. Therefore, issues in LTS can not be nominated for backport automatically. Also note that most bugs are not regressions (and are therefore present in current LTS line as well as in master), backporting all that would be lot more work and risk for LTS.

If I am not mistaken, the issue [was not](https://github.com/jenkinsci/jenkins/commit/6b350af44dee8e1e6cad4cce67941b3ebf88ac52) introduced in LTS release (all Daniel says is it appeared between 2 particular LTS release lines) but in 1.558 weekly release.

**Comment by TheFriendlyCoder** [comment]
@olivergondza 
> Unfortunately, there is no structural information in jira pointing to a version where the bug was spotted. Therefore, issues in LTS can not be nominated for backport automatically. 
Ah - of course. I totally missed that bit. That makes total sense.

> If I am not mistaken, the issue was not introduced in LTS release (all Daniel says is it appeared between 2 particular LTS release lines) but in 1.558 weekly release.
It is possible that the original code change that introduced the defect may have been committed first to the master and included in a weekly release.  All I can say for certain is that the bug in question does not exist in LTS v1.532.3 and it does exist in LTS v1.554.1 so *some* change that was included in the latter release has in fact caused this bug to appear on the LTS edition when it was not present before.

Given the debilitating nature of the bug this pull request is meant to correct (ie: it makes the tool absolutely unusable when building at scale using Visual Studio) hopefully it's resolution can be expedited as much as your process will allow.

**Comment by TheFriendlyCoder** [comment]
> For a fix to be accepted into an LTS release it generally needs to have soaked in the master branch for a few weeks as well (excluding security fixes).

Too bad this process didn't catch the defect associated with this pull request before being integrated into the LTS edition in the first place. :(

**Comment by oleg-nenashev** [comment]
@TheFriendlyCoder 
JENKINS-9104 is not a Jenkins bug at all. It's a specific behavior of an external tool, which breaks the common logic of the processes spawning in order to achieve some advantages (performance, etc.). developers of MVS stuff did not want to create an extra service process, so it's up to them.

In this PR we consider applying an **enhancement**, which allows to workaround the erroneous behavior of the tool. 

P.S: The issue is not in the large scale. It's caused by building in parallel on a single host. Jenkins provides pretty much features like Throttling, which allow to workaround the issue as well.

**Comment by TheFriendlyCoder** [comment]
> It's a specific behavior of an external tool, which breaks the common logic of the processes spawning in order to achieve some advantages (performance, etc.). developers of MVS stuff did not want to create an extra service process, so it's up to them.

Agreed. In fact, I think everyone involved in this discussion thread, including myself, would prefer to see Microsoft change their compiler to use a more appropriate thread-management design like a system service as you suggest. If you have any leverage with Microsoft then maybe you can convince them to make that change. From what I can tell the few bug reports I've found on the MS forums that suggest doing something like this are simply responded to with a "working as intended" reply. That, combined with the fact that Visual Studio has included this odd background process for several releases now, spread across many years leads me to believe that this behavior is not about to change any time soon.

That being the case, we are stuck between a rock and a hard place. We are not able to just up-and-drop support for Visual Studio at our firm, Microsoft appears unwilling to entertain suggestions to change this behavior in their compiler, and now all recent LTS editions (and master versions as well) of Jenkins are completely unusable for us. So as I see it this leaves us with the following 4 options:

1. stay on the old / outdated version of Jenkins indefinitely
2. fork Jenkins and make our own fix for this problem
3. get a fix / change / enhancement made to Jenkins then upgrade to the next version to include the change
4. Stop using Jenkins completely and migrate to another CI tool. 

We've chosen option #1 for a year or two now but this no longer a sustainable option. We need some improvements that only exist in newer editions, like support for SVN v1.8, which require us to move to a newer version. Option #2 would be a temporary solution at best since we're not in the business of building continuous integration tools and are not eager to have to support our own custom version indefinitely. I was hoping for option #3 since I'm certain others in the industry would find it of value as well. But maybe in the end we will have no choice but to consider option #4.

> JENKINS-9104 is not a Jenkins bug at all.
Semantics aside, all I know is that every LTS release up to v1.532.3 has worked perfectly well with Visual Studio, and every release afterwards does not work at all. Call it what you will, this fact is undeniable. In that spirit I am just hoping that 'something' can be done, sooner rather than later, to resolve this unfortunate state of affairs so that future LTS releases can become usable again.

> The issue is not in the large scale. It's caused by building in parallel on a single host.
I suppose that is an accurate statement, however my assumption was that smaller teams and smaller projects / codebases would have less of a need to build lots of things in parallel as we do (ie: the larger the codebase, the more you have to build. The more you have to build, the more benefit you get by doing as many operations in parallel as possible.) 

> Jenkins provides pretty much features like Throttling, which allow to workaround the issue as well.
I assume what you are suggesting is to use the job throttling plugin to prevent more than one compilation operation from running on any given agent at any one time, effectively serializing the use of the pdbsrv background process and avoiding the problem being described. Unfortunately this would have extremely detrimental effects on the performance of our builds.

To give you an example of what I am talking about, on a high-end development system (SSDs, i7 3.4Ghz 8-core, 32GB RAM, etc.) it takes about 10-12 hours to build one copy of our entire codebase from scratch when building the projects in series. Splitting as many of the compilation operations into parallel builds reduces this to about 4 hours or so. Basically, if we could no longer perform serialized compilation operations our build times would triple, again making it completely impractical to continue using Jenkins as our CI tool.

Again, while I do agree these problems are not totally isolated to large-scale development teams it is definitely exacerbated by large, complex build configurations like ours. And I'll also re-state it once again (because it never gets old :) ) that we are currently having no issues at all with the scale and complexity of our build configuration using Jenkins LTS v1.532.3, but we absolutely can not justify upgrading to any newer version because of the behavioral changes found in v1.554.1 and later.

**Comment by oleg-nenashev** [comment]
@TheFriendlyCoder 
Please be sure I totally support merging the change. The backporting is a more tough question, because new API definitely violates the common LTS approach.

Probably, you could just backport it to a custom branch and then build a custom core based on the LTS version. I've been using such approach for years to keep old instances running with the desired features.

Would it work for you? If yes, let me know if you need a help to build such custom version.

**Comment by TheFriendlyCoder** [comment]
> Please be sure I totally support merging the change.

Thanks for clarifying that. I was starting to have doubts.

> The backporting is a more tough question, because new API definitely violates the common LTS approach.

[small rant]
I wish this were true. In fact, one of the main reasons we are even considering upgrading our core version is because we need new versions of 1 or 2 specific plugins which do not support the core version we are currently on due to breaking API changes that have been made subsequent to v1.532.3. Therefore we *have* to first upgrade our core version before we can even consider upgrading the requisite plugins. :(
[end of rant]

Ranting aside, I took a quick look over the source changes in Daniel's pull request. Granted I am not a Java expert and have no knowledge of the internal workings of Jenkins, but it appears to me his changes should be relatively easy to incorporate into the core in a non-breaking way (Adding a new class, a couple of new methods on existing classes, some unit tests, etc.) thus eliminating your concerns regarding backporting.

Also, correct me if I'm mistaken but what I think may have been happening in the preceding discussion thread is consideration of adding additional changes to the original pull request to improve the integration with the rest of the Jenkins API / subsystems. That being the case could you not roll this 'enhancement' out in two phases: first implementing the new feature to support a process whitelist which could hopefully be backported to the LTS edition, then afterwards adjusting the implementation to improve its integration with the rest of the Jenkins API?

> Probably, you could just backport it to a custom branch and then build a custom core based on the LTS version. 

My main concern with that approach is whether it would make future upgrades more difficult or not. I would assume the maintenance strategy would be to select a future version that we are considering upgrading to and merging all changes made since our branch was cut to our custom branch and rebuilding. Assuming such merges are painless that might be manageable, I'm just not sure we can ensure that will be the case. :(

> Would it work for you?

I'd consider this a last resort. I'd prefer to exhaust all other possible solutions first though.

> If yes, let me know if you need a help to build such custom version.

That would be greatly appreciated if / when the time comes. Thank you.

**Comment by daniel-beck** [comment]
Given how many users appear to be affected by it, I'll try to assist @DanielWeber in getting his PR finished in time for the next LTS baseline release. If we manage to do that, it'll be available in the LTS release scheduled for mid/late August.

**Comment by addoula88** [comment]
Hello. What is the status of the whitelist feature?

**Comment by oleg-nenashev** [comment]
@addoula88 Extension point has been merged in https://github.com/jenkinsci/jenkins/pull/1684 . Now somebody needs to implement a plugin if it has not been created yet

**Pull request closed without merging**